File: src/main/java/com/thehecklers/theempire/TheEmpireApplication.java
package com.thehecklers.theempire;

import lombok.*;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.PostConstruct;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Random;

@SpringBootApplication
public class TheEmpireApplication {
    public static void main(String[] args) {
        SpringApplication.run(TheEmpireApplication.class, args);
    }
}

@Component
@AllArgsConstructor
class DataLoader {
    private final ShipRepository repo;

    @PostConstruct
    void loadShips() {
		List<String> ships = Arrays.asList("Enterprise", "Constitution", "Farragut", "Defiant", "Excalibur", "Exeter",
				"Lexington", "Hood", "Intrepid", "Voyager");
		List<String> captains = Arrays.asList("Adams", "Archer", "Decker", "Janeway", "Kirk", "Pike", "Sisko", "Spock",
				"Sulu", "Wesley");
		Random rnd = new Random();

    	repo.deleteAll();

		for (int x = 0; x < 1000; x++) {
			repo.save(new Ship(
				ships.get(rnd.nextInt(ships.size())),
				captains.get(rnd.nextInt(captains.size()))
			));
		}

		repo.findAll().forEach(System.out::println);
    }
}

@RestController
@AllArgsConstructor
class ShipController {
    private final ShipRepository repo;

    @GetMapping("/ships")
    Iterable<Ship> getAllShips() {
        return repo.findAll();
    }

    @GetMapping("/ships/{id}")
    Optional<Ship> getShipById(@PathVariable String id) {
        return repo.findById(id);
    }

    @GetMapping("/search")
    Iterable<Ship> getShipsCaptainedBy(@RequestParam(defaultValue = "Archer") String captain) {
        return repo.findShipByCaptain(captain);
    }
}

interface ShipRepository extends CrudRepository<Ship, String> {
    Iterable<Ship> findShipByCaptain(String captain);
}

@Data
@NoArgsConstructor
@RequiredArgsConstructor
@Document
class Ship {
    @Id
    private String id;
    @NonNull
    private String name, captain;
}

File: src/main/kotlin/com/thehecklers/theempire/TheEmpireApplication.kt
package com.thehecklers.theempire;

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.context.annotation.Bean
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.repository.reactive.ReactiveCrudRepository
import org.springframework.stereotype.Component
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.router
import reactor.core.publisher.Flux
import reactor.core.publisher.toFlux
import java.util.*
import javax.annotation.PostConstruct
import kotlin.random.Random

@SpringBootApplication
class TheEmpireApplication

fun main(args: Array<String>) {
    runApplication<TheEmpireApplication>(*args)
}

@Component
class DataLoader(private val repo: ShipRepository) {
    @PostConstruct
    fun loadShips() {
        val ships = listOf(
            "Ch'Tang",
            "Gr'oth",
            "Hegh'ta",
            "M'Char",
            "Maht-H'a",
            "Ning'tao",
            "Pagh",
            "T'Ong",
            "Vor'nak",
            "Ya'Vang"
        )
        val captains = listOf("Martok", "Koloth", "Kurn", "Kaybok", "Nu'Daq", "Lurkan", "Kargan", "K'Temoc", "Tanas")
        val rnd = Random

        repo.deleteAll().thenMany(
            (0..999).toFlux()
        )
            .map {
                Ship(
                    name = ships.get(rnd.nextInt(ships.size)),
                    captain = captains.get(rnd.nextInt(captains.size))
                )
            }
            .flatMap { repo.save(it) }
            .subscribe { println(it) }

    }
}

@Component
class ShipRouter(private val repo: ShipRepository) {
    @Bean
    fun shipRouting() = router {
        GET("/ships") { req -> ok().body(repo.findAll()) }
        GET("/ships/{id}") { req -> ok().body(repo.findById(req.pathVariable("id"))) }
        // How to add @RequestParam(defaultValue = "Martok") to following handler ???
        GET("/search") { req -> ok().body(repo.findShipByCaptain(req.queryParam("captain"))) }
    }
}

interface ShipRepository : ReactiveCrudRepository<Ship, String> {
    fun findShipByCaptain(captain: Optional<String>): Flux<Ship>
}

@Document
data class Ship(@Id val id: String? = null, val name: String, val captain: String)